src/menu/charprofile.c

```diff
@@ -0,0 +1,395 @@
+	r_mat_mv_push();
+	r_mat_mv_push();
+
+	if(e->drawdata != 0) {
+		r_mat_mv_translate(0, -300 * e->drawdata, 0);
+		r_mat_mv_rotate(M_PI * e->drawdata, 1, 0, 0);
+	}
+
+	text_draw(profiles[selected].fullname, &(TextParams) {
+		.align = ALIGN_CENTER,
+		.font = "big",
+		.shader = "text_default",
+		.color = RGBA(o, o, o, o),
+	});
+	r_mat_mv_pop();
+
+	if(e->drawdata) {
+		o = 1 - e->drawdata * 3;
+	} else {
+		o = 1;
+	}
+
+	text_draw(profiles[selected].title, &(TextParams) {
+		.align = ALIGN_CENTER,
+		.pos = { 20*(1-o), 30 },
+		.shader = "text_default",
+		.color = RGBA(o, o, o, o),
+	});
+	r_mat_mv_pop();
+
+	text_draw_wrapped(profiles[selected].description, DESCRIPTION_WIDTH, &(TextParams) {
+		.align = ALIGN_LEFT,
+		.pos = { -175, 120 },
+		.font = "small",
+		.shader = "text_default",
+		.color = RGBA(o, o, o, o),
+	});
+	r_mat_mv_pop();
+
+	o = 0.3*sin(m->frames/20.0)+0.5;
+	r_shader("sprite_default");
+
+	r_draw_sprite(&(SpriteParams) {
+		.sprite = "menu/arrow",
+		.pos = { 30, SCREEN_H/3+10 },
+		.color = RGBA(o, o, o, o),
+		.scale = { 0.5, 0.7 },
+	});
+
+	r_draw_sprite(&(SpriteParams) {
+		.sprite = "menu/arrow",
+		.pos = { 30 + 340, SCREEN_H/3+10 },
+		.color = RGBA(o, o, o, o),
+		.scale = { 0.5, 0.7 },
+		.flip.x = true,
+	});
+
+	r_state_pop();
+}
+
+static void action_show_character(MenuData *m, void *arg) {
+	return;
+}
+
+static void add_character(MenuData *m, int i) {
+	if(strcmp(profiles[i].name, "locked")) {
+		log_debug("adding character: %s", profiles[i].name);
+		Sprite *spr = portrait_get_base_sprite(profiles[i].name, NULL);
+		MenuEntry *e = add_menu_entry(m, NULL, action_show_character, spr);
+		e->transition = NULL;
+	}
+}
+
+static void charprofile_free(MenuData *m) {
+	free(m->context);
+}
+
+// TODO: add a better drawing animation for character selection
+static bool charprofile_input_handler(SDL_Event *event, void *arg) {
+	MenuData *m = arg;
+	CharProfileContext *ctx = m->context;
+	TaiseiEvent type = TAISEI_EVENT(event->type);
+
+	if(type == TE_MENU_CURSOR_LEFT) {
+		play_sfx_ui("generic_shot");
+		m->cursor--;
+		ctx->face = 0;
+	} else if(type == TE_MENU_CURSOR_RIGHT) {
+		play_sfx_ui("generic_shot");
+		m->cursor++;
+		ctx->face = 0;
+	} else if(type == TE_MENU_ACCEPT) {
+		if(check_unlocked_profile(m->cursor) != PROFILE_LOCKED) {
+			play_sfx_ui("generic_shot");
+			// show different expressions for selected character
+			ctx->face++;
+			if(!profiles[m->cursor].faces[ctx->face]) {
+				ctx->face = 0;
+			}
+		}
+	} else if(type == TE_MENU_ABORT) {
+		play_sfx_ui("hit");
+		close_menu(m);
+	} else {
+		return false;
+	}
+
+	m->cursor = (m->cursor % m->entries.num_elements) + m->entries.num_elements * (m->cursor < 0);
+
+	return false;
+
+}
+
+void preload_charprofile_menu(void) {
+	for(int i = 0; i < NUM_PROFILES-1; i++) {
+		portrait_preload_base_sprite(profiles[i].name, NULL, RESF_PERMANENT);
+		preload_resource(RES_TEXTURE, profiles[i].background, RESF_PERMANENT);
+	}
+};
```

Comment:
Actually, one more thing: the faces are not preloaded (call `portrait_preload_face_sprite` for each).